### PR TITLE
feat: add weights to provisioners and order provisioner scheduling by weights

### LINKED ISSUE #2024: weight provisioners?
**What karpenter features are relevant?**
provisioners

**How should the docs be improved?**
We currently deploy a single provisioner with both X64 and ARMv8 archs
```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
...
  generation: 1
  name: default-dev-ue1-008-flying-possum
  resourceVersion: "4109"
spec:
  limits:
    resources:
      cpu: 1k
  provider:
    amiFamily: AL2
    blockDeviceMappings:
    - deviceName: /dev/xvda
      ebs:
        deleteOnTermination: true
        iops: 10000
        throughput: 125
        volumeSize: 100Gi
        volumeType: gp3
    securityGroupSelector:
      aws-ids: sg-xxxx
    subnetSelector:
      aws-ids: subnet-xxx
    tags:
      Environment: dev
  requirements:
  - key: topology.kubernetes.io/zone
    operator: In
    values:
    - us-east-1a
    - us-east-1b
    - us-east-1c
    - us-east-1d
    - us-east-1f
  - key: karpenter.sh/capacity-type
    operator: In
    values:
    - spot
    - on-demand
  - key: node.kubernetes.io/instance-type
    operator: In
    values:
    - c6a.large
    - c6g.large
    - c6i.large
  - key: kubernetes.io/arch
    operator: In
    values:
    - amd64
    - arm64
  ttlSecondsAfterEmpty: 120
  ttlSecondsUntilExpired: 2592000
status:
  resources:
    attachable-volumes-aws-ebs: "39"
    cpu: "2"
    ephemeral-storage: 113224684Ki
    memory: 3915268Ki
    pods: "29"
```

Our question is if there is an option to have two provisioners, one per arch but with different weights in each.
We want to offer both arch but not all workloads work on ARM, but we can't currently modify the Deployment for all of them to pin an arch.
If we could influence Karpenter to pick x64 over ARMv8 (it tends to pick graviton cause it is cheaper, I believe), and then specific workloads that pick arm arch would ask for those specific nodes.

Thanks

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by jonathan-innis**
Fixes #2024 

**Description**

- Add `.spec.weight` to the `v1alpha5/Provisioner` spec
- By sorting provisioners before scheduling, nodeTemplates used for scheduling will try to use these provisioners first

**How was this change tested?**

* `make test`
* Manual validation with multiple provisioners

**Does this change impact docs?**
- [x] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [ ] No

**Release Note**

```release-note
None
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/apis/provisioning/v1alpha5/provisioner.go [hunk 1]
```diff
@@ -70,6 +72,12 @@ type ProvisionerSpec struct {
 	TTLSecondsUntilExpired *int64 `json:"ttlSecondsUntilExpired,omitempty"`
 	// Limits define a set of bounds for provisioning capacity.
 	Limits *Limits `json:"limits,omitempty"`
+	// Weight is the priority given to the provisioner during scheduling. A higher
+	// numerical weight indicates that this provisioner will be ordered
+	// ahead of other provisioners with lower weights. A provisioner with no weight
+	// will be treated as if it is a provisioner with a weight of 0.
+	// +optional
+	Weight int `json:"weight,omitempty"`
```
[the remark this answers is not in the document]
Comment (jonathan-innis) [reply pkg/apis/provisioning/v1alpha5/provisioner.go:80]: This sets the default as 1 and doesn't omit either. Sure, makes sense to take the same approach as [WeightedPodAffinityTerm](https://pkg.go.dev/k8s.io/api/core/v1#WeightedPodAffinityTerm)

[the remark this answers is not in the document]
Comment (ellistarn) [reply pkg/apis/provisioning/v1alpha5/provisioner.go:80]: Related: https://karpenter.sh/v0.13.2/contributing/design-guide/#does-your-change-introduce-new-apis
> Rely on existing concepts and idioms from the Kubernetes ecosystem. Look to [Kubernetes APIs](https://pkg.go.dev/k8s.io/api/core/v1) and projects like [Tekton](https://github.com/tektoncd/cli), [Knative](https://github.com/knative/serving), and [ACK](https://github.com/aws-controllers-k8s) and find concepts that will be familiar to users.

File: pkg/apis/provisioning/v1alpha5/provisioner.go [hunk 2]
```diff
@@ -132,6 +134,12 @@ type ProvisionerList struct {
 // by their priority weight in-place
 func (pl *ProvisionerList) OrderByWeight() {
 	sort.Slice(pl.Items, func(a, b int) bool {
-		return pl.Items[a].Spec.Weight > pl.Items[b].Spec.Weight
+		if pl.Items[b].Spec.Weight == nil {
```
Comment (ellistarn) [inline pkg/apis/provisioning/v1alpha5/provisioner.go:137]: Can `ptr.Int32(pl.Items[b].Spec.Weight)` simplify these if checks? It converts nil -> 0, which is the semantic we want.

**Review by bwagner5 — DISMISSED** [review]
lgtm!

**Review by bwagner5 — APPROVED** [review]
lgtm

**Review by ellistarn — APPROVED** [review]
Nicely done!

This change needs some docs for "Configuring global scheduling preferences"

**Comment by jonathan-innis** [comment]
Will post follow-up PR for docs changes

**Pull request merged**